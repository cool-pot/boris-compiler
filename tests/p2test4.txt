*** more fun with simple variables

defun fib(x)
  if x<0
    then print 9999; return 0-1;
  elsif x==0 then
    return 0;
  elsif x==1 then
    return 1;
  else *** x>1
    *** do we need a declaration for these? wait until next project stage!
    a,b=0,1;
    count = 1;
    while count <= x do *** off by one error?
      a,b = b,a+b; *** make sure this is OK in the sematics of the language
    end while
    return a;
  end if
end defun

*** test the function
foreach c in 1 .. 10 do
  print c;
  print fib c;
end for

*** now recursively
defun fib_two (x)
  if x<0
    then print 9999; return 0-1;
  elsif x==0 then
    return 0;
  elsif x==1 then
    return 1;
  else *** x>1
    return fib_two(x-1)+fib_two(x-2);
  end if
end defun

*** test that
foreach c in 1 .. 10 do
  print c;
  print fib_two c;
end for
