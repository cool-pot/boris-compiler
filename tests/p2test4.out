> Hello from parserTester for boris compiler!
Start parse tree visualization:

[ROOT_INPUT]
    [SDD_LIST]
        [DEFN_AS_SDD]
            [FUNC_DEFN]
                [PLACEHOLDER-262]defun
                [ID]fib
                [PLACEHOLDER-291](
                [ID]x
                [EMPTY_NODE]
                [PLACEHOLDER-291](
                [BODY]
                    [SD_LIST]
                        [STATEMENT_AS_SD]
                            [IF_ELSE_STATEMENT]
                                [PLACEHOLDER-266]if
                                [BOOLEXPR]
                                    [SINGLE_ID_AS_EXPR]
                                        [ID]x
                                    [PLACEHOLDER-278]<
                                    [SINGLE_INT_AS_EXPR]
                                        [INT]0
                                [PLACEHOLDER-267]then
                                [STATEMENT_LIST]
                                    [PRINT_STATEMENT]
                                        [PLACEHOLDER-274]print
                                        [SINGLE_INT_AS_EXPR]
                                            [INT]9999
                                        [PLACEHOLDER-290];
                                    [STATEMENT_LIST]
                                        [RETURN_STATEMENT]
                                            [PLACEHOLDER-273]return
                                            [EXPR_MINUS_EXPR]
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]0
                                                [PLACEHOLDER-285]-
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]1
                                            [PLACEHOLDER-290];
                                        [EMPTY_NODE]
                                [ELSIF_SENTENCE_LIST]
                                    [ELSIF_SENTENCE]
                                        [PLACEHOLDER-268]elsif
                                        [BOOLEXPR]
                                            [SINGLE_ID_AS_EXPR]
                                                [ID]x
                                            [PLACEHOLDER-282]==
                                            [SINGLE_INT_AS_EXPR]
                                                [INT]0
                                        [PLACEHOLDER-267]then
                                        [STATEMENT_LIST]
                                            [RETURN_STATEMENT]
                                                [PLACEHOLDER-273]return
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]0
                                                [PLACEHOLDER-290];
                                            [EMPTY_NODE]
                                    [ELSIF_SENTENCE_LIST]
                                        [ELSIF_SENTENCE]
                                            [PLACEHOLDER-268]elsif
                                            [BOOLEXPR]
                                                [SINGLE_ID_AS_EXPR]
                                                    [ID]x
                                                [PLACEHOLDER-282]==
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]1
                                            [PLACEHOLDER-267]then
                                            [STATEMENT_LIST]
                                                [RETURN_STATEMENT]
                                                    [PLACEHOLDER-273]return
                                                    [SINGLE_INT_AS_EXPR]
                                                        [INT]1
                                                    [PLACEHOLDER-290];
                                                [EMPTY_NODE]
                                        [EMPTY_NODE]
                                [ELSE_SENTENCE]
                                    [PLACEHOLDER-269]else
                                    [STATEMENT_LIST]
                                        [LHS_ASSIGN_EXPR_AS_STATEMENT]
                                            [LHS]
                                                [SINGLE_ID_AS_LHSITEM]
                                                    [ID]a
                                                [COMMA_LHSITEN_LIST]
                                                    [PLACEHOLDER-276],
                                                    [SINGLE_ID_AS_LHSITEM]
                                                        [ID]b
                                                    [EMPTY_NODE]
                                            [PLACEHOLDER-293]=
                                            [EXPR_COMMA_EXPR]
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]0
                                                [PLACEHOLDER-276],
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]1
                                            [PLACEHOLDER-290];
                                        [STATEMENT_LIST]
                                            [LHS_ASSIGN_EXPR_AS_STATEMENT]
                                                [LHS]
                                                    [SINGLE_ID_AS_LHSITEM]
                                                        [ID]count
                                                    [EMPTY_NODE]
                                                [PLACEHOLDER-293]=
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]1
                                                [PLACEHOLDER-290];
                                            [STATEMENT_LIST]
                                                [WHILE_STATEMENT]
                                                    [PLACEHOLDER-264]while
                                                    [BOOLEXPR]
                                                        [SINGLE_ID_AS_EXPR]
                                                            [ID]count
                                                        [PLACEHOLDER-280]<=
                                                        [SINGLE_ID_AS_EXPR]
                                                            [ID]x
                                                    [PLACEHOLDER-265]do
                                                    [STATEMENT_LIST]
                                                        [LHS_ASSIGN_EXPR_AS_STATEMENT]
                                                            [LHS]
                                                                [SINGLE_ID_AS_LHSITEM]
                                                                    [ID]a
                                                                [COMMA_LHSITEN_LIST]
                                                                    [PLACEHOLDER-276],
                                                                    [SINGLE_ID_AS_LHSITEM]
                                                                        [ID]b
                                                                    [EMPTY_NODE]
                                                            [PLACEHOLDER-293]=
                                                            [EXPR_COMMA_EXPR]
                                                                [SINGLE_ID_AS_EXPR]
                                                                    [ID]b
                                                                [PLACEHOLDER-276],
                                                                [EXPR_PLUS_EXPR]
                                                                    [SINGLE_ID_AS_EXPR]
                                                                        [ID]a
                                                                    [PLACEHOLDER-284]+
                                                                    [SINGLE_ID_AS_EXPR]
                                                                        [ID]b
                                                            [PLACEHOLDER-290];
                                                        [EMPTY_NODE]
                                                    [PLACEHOLDER-263]end
                                                    [PLACEHOLDER-264]while
                                                [STATEMENT_LIST]
                                                    [RETURN_STATEMENT]
                                                        [PLACEHOLDER-273]return
                                                        [SINGLE_ID_AS_EXPR]
                                                            [ID]a
                                                        [PLACEHOLDER-290];
                                                    [EMPTY_NODE]
                                [PLACEHOLDER-263]end
                                [PLACEHOLDER-266]if
                        [EMPTY_NODE]
                [PLACEHOLDER-263]end
                [PLACEHOLDER-262]defun
        [SDD_LIST]
            [STATEMENT_AS_SDD]
                [FOREACH_RANGE_STATEMENT]
                    [PLACEHOLDER-270]foreach
                    [ID]c
                    [PLACEHOLDER-272]in
                    [RANGE]
                        [SINGLE_INT_AS_EXPR]
                            [INT]1
                        [PLACEHOLDER-275]..
                        [SINGLE_INT_AS_EXPR]
                            [INT]10
                    [PLACEHOLDER-265]do
                    [STATEMENT_LIST]
                        [PRINT_STATEMENT]
                            [PLACEHOLDER-274]print
                            [SINGLE_ID_AS_EXPR]
                                [ID]c
                            [PLACEHOLDER-290];
                        [STATEMENT_LIST]
                            [PRINT_STATEMENT]
                                [PLACEHOLDER-274]print
                                [FUNC_CALL_AS_EXPR]
                                    [ID]fib
                                    [SINGLE_ID_AS_EXPR]
                                        [ID]c
                                [PLACEHOLDER-290];
                            [EMPTY_NODE]
                    [PLACEHOLDER-263]end
                    [PLACEHOLDER-271]for
            [SDD_LIST]
                [DEFN_AS_SDD]
                    [FUNC_DEFN]
                        [PLACEHOLDER-262]defun
                        [ID]fib_two
                        [PLACEHOLDER-291](
                        [ID]x
                        [EMPTY_NODE]
                        [PLACEHOLDER-291](
                        [BODY]
                            [SD_LIST]
                                [STATEMENT_AS_SD]
                                    [IF_ELSE_STATEMENT]
                                        [PLACEHOLDER-266]if
                                        [BOOLEXPR]
                                            [SINGLE_ID_AS_EXPR]
                                                [ID]x
                                            [PLACEHOLDER-278]<
                                            [SINGLE_INT_AS_EXPR]
                                                [INT]0
                                        [PLACEHOLDER-267]then
                                        [STATEMENT_LIST]
                                            [PRINT_STATEMENT]
                                                [PLACEHOLDER-274]print
                                                [SINGLE_INT_AS_EXPR]
                                                    [INT]9999
                                                [PLACEHOLDER-290];
                                            [STATEMENT_LIST]
                                                [RETURN_STATEMENT]
                                                    [PLACEHOLDER-273]return
                                                    [EXPR_MINUS_EXPR]
                                                        [SINGLE_INT_AS_EXPR]
                                                            [INT]0
                                                        [PLACEHOLDER-285]-
                                                        [SINGLE_INT_AS_EXPR]
                                                            [INT]1
                                                    [PLACEHOLDER-290];
                                                [EMPTY_NODE]
                                        [ELSIF_SENTENCE_LIST]
                                            [ELSIF_SENTENCE]
                                                [PLACEHOLDER-268]elsif
                                                [BOOLEXPR]
                                                    [SINGLE_ID_AS_EXPR]
                                                        [ID]x
                                                    [PLACEHOLDER-282]==
                                                    [SINGLE_INT_AS_EXPR]
                                                        [INT]0
                                                [PLACEHOLDER-267]then
                                                [STATEMENT_LIST]
                                                    [RETURN_STATEMENT]
                                                        [PLACEHOLDER-273]return
                                                        [SINGLE_INT_AS_EXPR]
                                                            [INT]0
                                                        [PLACEHOLDER-290];
                                                    [EMPTY_NODE]
                                            [ELSIF_SENTENCE_LIST]
                                                [ELSIF_SENTENCE]
                                                    [PLACEHOLDER-268]elsif
                                                    [BOOLEXPR]
                                                        [SINGLE_ID_AS_EXPR]
                                                            [ID]x
                                                        [PLACEHOLDER-282]==
                                                        [SINGLE_INT_AS_EXPR]
                                                            [INT]1
                                                    [PLACEHOLDER-267]then
                                                    [STATEMENT_LIST]
                                                        [RETURN_STATEMENT]
                                                            [PLACEHOLDER-273]return
                                                            [SINGLE_INT_AS_EXPR]
                                                                [INT]1
                                                            [PLACEHOLDER-290];
                                                        [EMPTY_NODE]
                                                [EMPTY_NODE]
                                        [ELSE_SENTENCE]
                                            [PLACEHOLDER-269]else
                                            [STATEMENT_LIST]
                                                [RETURN_STATEMENT]
                                                    [PLACEHOLDER-273]return
                                                    [EXPR_PLUS_EXPR]
                                                        [FUNC_CALL_AS_EXPR]
                                                            [ID]fib_two
                                                            [LPAR_EXPR_LPAR]
                                                                [PLACEHOLDER-291](
                                                                [EXPR_MINUS_EXPR]
                                                                    [SINGLE_ID_AS_EXPR]
                                                                        [ID]x
                                                                    [PLACEHOLDER-285]-
                                                                    [SINGLE_INT_AS_EXPR]
                                                                        [INT]1
                                                                [PLACEHOLDER-292])
                                                        [PLACEHOLDER-284]+
                                                        [FUNC_CALL_AS_EXPR]
                                                            [ID]fib_two
                                                            [LPAR_EXPR_LPAR]
                                                                [PLACEHOLDER-291](
                                                                [EXPR_MINUS_EXPR]
                                                                    [SINGLE_ID_AS_EXPR]
                                                                        [ID]x
                                                                    [PLACEHOLDER-285]-
                                                                    [SINGLE_INT_AS_EXPR]
                                                                        [INT]2
                                                                [PLACEHOLDER-292])
                                                    [PLACEHOLDER-290];
                                                [EMPTY_NODE]
                                        [PLACEHOLDER-263]end
                                        [PLACEHOLDER-266]if
                                [EMPTY_NODE]
                        [PLACEHOLDER-263]end
                        [PLACEHOLDER-262]defun
                [SDD_LIST]
                    [STATEMENT_AS_SDD]
                        [FOREACH_RANGE_STATEMENT]
                            [PLACEHOLDER-270]foreach
                            [ID]c
                            [PLACEHOLDER-272]in
                            [RANGE]
                                [SINGLE_INT_AS_EXPR]
                                    [INT]1
                                [PLACEHOLDER-275]..
                                [SINGLE_INT_AS_EXPR]
                                    [INT]10
                            [PLACEHOLDER-265]do
                            [STATEMENT_LIST]
                                [PRINT_STATEMENT]
                                    [PLACEHOLDER-274]print
                                    [SINGLE_ID_AS_EXPR]
                                        [ID]c
                                    [PLACEHOLDER-290];
                                [STATEMENT_LIST]
                                    [PRINT_STATEMENT]
                                        [PLACEHOLDER-274]print
                                        [FUNC_CALL_AS_EXPR]
                                            [ID]fib_two
                                            [SINGLE_ID_AS_EXPR]
                                                [ID]c
                                        [PLACEHOLDER-290];
                                    [EMPTY_NODE]
                            [PLACEHOLDER-263]end
                            [PLACEHOLDER-271]for
                    [EMPTY_NODE]
